test/components/syncTest.js

```diff
@@ -0,0 +1,788 @@
+          })
+        })
+    })
+
+    it('sync categories can be enabled', function * () {
+      yield this.app.client
+        .tabByIndex(0)
+        .loadUrl(prefsUrl)
+        .waitForVisible(syncTab)
+        .click(syncTab)
+        .waitForVisible(syncSwitch)
+        .click(syncSwitch)
+        .waitForExist('#syncData .switchBackground')
+        .click('#syncData .switchBackground')
+        .windowByUrl(Brave.browserWindowUrl)
+        .waitUntil(function () {
+          return this.getAppState().then((val) => {
+            return val.value.settings['sync.type.bookmark'] === false
+          })
+        })
+    })
+
+    it('shows sync QR code and words', function * () {
+      yield this.app.client
+        .tabByIndex(0)
+        .loadUrl(prefsUrl)
+        .waitForVisible(syncTab)
+        .click(syncTab)
+        .waitForVisible(syncSwitch)
+        .click(syncSwitch)
+        .waitForExist(newDeviceButton)
+        .click(newDeviceButton)
+        .click('[data-l10n-id="syncShowPassphrase"]')
+        .waitUntil(function () {
+          return this.getText('#syncPassphrase').then((text) => {
+            return text === 'a a a a\na a a a\na a a a\na a a a'
+          })
+        })
+        .click('[data-l10n-id="syncShowQR"]')
+        .waitUntil(function () {
+          return this.getAttribute('#syncQR', 'src').then((text) => {
+            return text === 'data:image/png;base64,foo'
+          })
+        })
+    })
+  })
+})
+
+describe('Syncing bookmarks', function () {
+  Brave.beforeAllServerSetup(this)
+
+  function * setup (seed) {
+    yield Brave.startApp()
+    yield setupBrave(Brave.app.client)
+    yield setupSync(Brave.app.client, seed)
+  }
+
+  before(function * () {
+    this.page1Url = Brave.server.url('page1.html')
+    this.page1Title = 'Page 1'
+    this.page2Url = Brave.server.url('page2.html')
+    this.page2Title = 'Page 2 - to be Renamed'
+    this.page2TitleUpdated = 'All shall hail pyramids'
+    this.page3Url = Brave.server.url('page3.html')
+    this.page3Title = 'Page 3 - to be Deleted'
+    this.folder1Title = 'Folder 1'
+    this.folder1Page1Url = Brave.server.url('page4.html')
+    this.folder1Page1Title = 'Page 1.4'
+    this.folder1Page2Url = Brave.server.url('page5.html')
+    this.folder1Page2Title = 'Page 1.5'
+    this.folder2Title = 'Folder 2 - to be Deleted'
+    this.folder2Page1Url = Brave.server.url('page6.html')
+    this.folder2Page1Title = 'Page - to be Deleted'
+    this.seed = new Immutable.List(crypto.randomBytes(32))
+    yield setup(this.seed)
+
+    // For Create: Bookmark page 1
+    yield bookmarkUrl(this.page1Url, this.page1Title)
+
+    // For Update: Bookmark page 2 and rename it
+    yield Brave.app.client
+      .tabByIndex(0)
+      .waitForUrl(this.page1Url)
+    yield bookmarkUrl(this.page2Url, this.page2Title)
+    yield Brave.app.client
+      .pause(1000) // XXX: Helps to correctly order Create and Update
```

Comment:
in general i agree on removing explicit pauses; if i understand correctly, it is tricky here because this is waiting on the server to receive the records and process them, which is not something that can be detected client-side (ex: using a waitUntil for some appState change). since the server can't push notifications to the client, we would have to have the client poll the server periodically to detect when a sync operation has finished.